Constrain Jacobians at the element level only.

roystgnr opened the pull request:
## Reason
This should fix https://github.com/libMesh/libmesh/issues/3504 for MOOSE.

## Design
Apply Jacobian constraints consistently at the element level, then don't apply them redundantly at the global level.

## Impact
No API changes.  This also removes the post-facto constraint enforcement sweep we previously did in between residual and Jacobian evaluations, so slight changes in convergence behavior might be possible for problems with strongly-enforced constraints (AMR, PBCs, IGA), which should also run much, much, much faster.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1473846/) on ed6a112 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24145/site)

This comment will be updated on new commits.

roystgnr commented:
I can repeat the failures here, and to some extent it looks like the trouble is that the new system is working *too* well.  `rel diff:  1.3983984e-06 ~  1.3986524e-06 = 1.81555e-04` is just what we'd expect to see from a problem with an initial constraint misfit and a tolerance around 1e-10.  Then in the next time step we're at 1e-11, and in time steps after that we're not seeing any exodiffs.

Maybe I need to break up the constrain_in_solver option more finely.  By default it could apply only the solution postprocessing step, which would still skip the expensive Jacobian postprocessing step and would match well enough to keep exodiff happy; with an additional option it would *then* skip every step and give us diffs on the order of our solver tolerances.

lindsayad commented:
I would lean towards regolding instead if the new solution is equally valid, as opposed to adding code complexity for no user-based reason

roystgnr commented:
> I would lean towards regolding instead if the new solution is equally valid,

IMHO it is, in the periodic BC cases with O(1e-11).  We're still getting the level of accuracy we requested from the solver, the residual is just distributed in a different way and our exodiff tolerances are too tight for it.

On the other hand, the `MultiPhase.derivativetwophasematerial` test in `modules/phase_field` isn't just converging to something O(1e-11) off, it's failing to converge.

If I play with solver settings, I can fix that.  Bump the max nonlinear iterations from 10 to 21 and we converge again, to the same result within exodiff tolerances.

I'm going to experiment a bit more with continuing to clean up our constraints at the solver level.  I know I didn't get a *complete* summary of the discussions of convergence issues last week, but I'm guessing "Boy I wish some of my solves took twice as many steps" wasn't on the list.

I'm also going to see if I can figure out what's going on exactly here.  If I comment out the new `petsc_solver->set_exact_constraint_enforcement(false);` then I get back to a solve in 3 nonlinear steps.  If I run with `-snes_mf_operator -snes_fd -mat_fd_type ds` then I **also** get back to a solve in 3 nonlinear steps.  Maybe there's some place where we're incorrectly constraining (or more-incorrectly failing to constrain) an element matrix, but we were getting away with it because of the global Jacobian constraint application?

lindsayad commented:
Yea I would see what `-snes_test_jacobian` says before and after this PR

roystgnr commented:
We are doing something wrong in the multiphysics case.  With three variables, it's as if constraint equation rows are getting added once per containing_element to the Jacobian and once-per-containing_element-per-variable to the residual...

lindsayad commented:
Is that using AD objects? I'm wondering if I goofed something

roystgnr commented:
> Is that using AD objects?

I think it's a Cahn-Allen case?  I'd be surprised if that used AD.

> I'm wondering if I goofed something

Nah.  If there was a screwup in AD then adding the global constraint application wouldn't be enough to fix it, since that only touches the constrained rows.

I didn't properly consider the case where constraints are being applied to *submatrices* of an element Jacobian and *subvectors* of an element residual, so the bug is probably on my end.  MOOSE needs to do that, that's the only way we can efficiently run crazy physics with a thousand variables and sparse coupling between them, but it's also possible we're doing something wrong with the constraint applications on the MOOSE side.

I've been kind of stymied trying to debug with even this small test case but if I'm right about the root problem then it ought to be easy enough to replicate in a microscopic test case; I'll do that next.

roystgnr commented:
> but it's also possible we're doing something wrong with the constraint applications on the MOOSE side.

Ha!  Score one for desperate-hope-I-got-libMesh-right.  `_residual_vector_tags` has `"NONTIME"`, `"RESIDUAL"`, and `"TIME"` entries, we end up calling constraints on each of them (which is correct), and in each case we add the residual terms for constrained DoFs (which is incorrect).  Then presumably we sum them in later and so there we get 3x the constraint-residual terms.  The fact that it was a system of 3 variables that was most sensitive to this was a red herring.

roystgnr commented:
That said ... this is frightening from a performance and modularity as well as a correctness standpoint.  Why are we doing an expensive residual constraint on every entry in a sum, rather than just once on the summed values?  Once I fix the redundancy here, are there any of the other dozens of processResidual/processResiduals/processLocalResidual calls which might be making the same mistake?

lindsayad commented:
`Assembly` should be where we're doing all of this and we don't really provide any virtual interfaces there

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -6336,6 +6349,21 @@ FEProblemBase::computeResidualTags(const std::set<TagID> & tags)
+  // safely add heterogeneous terms anywhere there.  So we'll apply
+  // those terms once, at the global level.
+  {
+    auto & residual = _current_nl_sys->getVector(_current_nl_sys->residualVectorTag());
```
We should assert that this tag is in `tags`

roystgnr commented:
> By default it could apply only the solution postprocessing step

This is what we have to do, though I'll see if I can figure out how to do it properly at the MOOSE level rather than the libMesh level.  Digging down to the minimal case finally made the remaining bug obvious.  An implicit-Euler residual for a `LAGRANGE` u on a single `EDGE2` on [0,6] has element residual `r = [2*u0+u1; u0+2*u1]` with element Jacobian `J = [2 1; 1 2]`.  If we have the periodic constraint `u1 = u0`, that `x = C u` equation corresponds to the constraint matrix `C = [1 0; 1 0]`.  Constraining a residual vector with this gives `C^T r = [3*(u0+u1); 0]` ... but that residual *doesn't* have a Jacobian restricted to the 0-0 "block", the way `C^T J C = [6 0; 0 0]` does, because we didn't *apply* the `x = C u` restriction before evaluating `r(x)`, and dR/d**u** doesn't have that same dR/d**x** block structure.

roystgnr commented:
> This is what we have to do

... and even this is tricky - if we do pre-enforce constraints, we lose the information about their discrepancies, and we then can't evaluate the terms we want in the residual to control those discrepancies.  This was only working in libMesh because we were using additional vectors at that level.  With a sufficiently modular API we could have pre-constrained just the element solution vector, but we can't expect that a MOOSE kernel is actually relying solely on prepopulated values rather than querying `current_local_solution` etc. itself.

Maybe it's time to throw in the towel on both ends, continue relying on global constraint enforcement, except calculate element constraints on the Jacobian anyway so as to make sure the global sparsity pattern doesn't end up shrunk away...

lindsayad commented:
> MOOSE kernel is actually relying solely on prepopulated values rather than querying `current_local_solution` etc. itself.

~MOOSE kernels in general shouldn't be relying on querying the current solution vector directly. There is at least one layer of abstraction (the variable classes) and more often two (coupleable interface) that kernels should be going through. In rare cases there are residual objects that query the current solution vector~ I realized that my on-the-fly evaluation system (mostly for finite volume) always queries the current solution vector directly :disappointed:

lindsayad commented:
> I realized that my on-the-fly evaluation system (mostly for finite volume) always queries the current solution vector directly disappointed

But I do think that we could change this

> Maybe it's time to throw in the towel on both ends, continue relying on global constraint enforcement, except calculate element constraints on the Jacobian anyway so as to make sure the global sparsity pattern doesn't end up shrunk away

This seems like a very sad conclusion. Maybe we need to go to the source of the problem and do something like https://gitlab.com/petsc/petsc/-/issues/852 ?

roystgnr commented:
On the bright side, I got an FD-consistent Jacobian *before* the segfault!

roystgnr commented:
And I didn't screw anything up with the rebase; I still get correct manual-vs-FD Jacobians with phase_field.  Hopefully whatever's not there in the framework tests is simple; it looks almost like the opposite of the problem I fixed for phase_field...

roystgnr commented:
And it is.  Now we're getting multiple constraint row additions to element Jacobians because of hanging node valence (which didn't apply to the periodic BCs constraints I was testing with), but we're not getting the same number of additions to element residuals because I gave up on adding those at the element residual level.

roystgnr committed:
Apply asymmetric constraints to element Jacobians

This won't match what we're doing in the residual, but it will get
entries into the otherwise unused bits of the sparsity pattern, which
should prevent PETSc from condensing those entries away and thereby let
the PetscNonlinearSolver global constraint enforcement work quickly.

This should fix libMesh/libmesh#3504 for MOOSE.

roystgnr commented:
Finally I got something correct and sound in theory, even if it wasn't the optimal solution ... but in practice this seems to actually be running *slower* for me, not faster, and I do not yet see why.

roystgnr commented:
I'll profile this weekend.

lindsayad commented:
Should we try to revive this?

roystgnr commented:
It wouldn't be crazy.  At the very least we could make it an option for a while until we figure out why some tests show significant differences.
